File: src/chain-observers/providers/custom-operation-provider.ts
import { custom, custom_json } from "@hiveio/wax";
import { TRegisterEvaluationContext } from "../classifiers/collector-classifier-base";
import { IOperationTransactionPair, OperationClassifier } from "../classifiers/operation-classifier";
import { TProviderEvaluationContext } from "../factories/data-evaluation-context";
import { ProviderBase } from "./provider-base";

export type TCustomOperationProvided<TOperationId extends Array<string | number>> = {
  [K in TOperationId[number]]: Array<IOperationTransactionPair<custom_json | custom>>;
};

export interface ICustomOperationProviderData<TOperationId extends Array<string | number>> {
  customOperations: Partial<TCustomOperationProvided<TOperationId>>;
};

export interface ICustomOperationProviderOptions {
  ids: Array<string | number>;
}

export class CustomOperationProvider<
  TOperationId extends Array<string | number> = Array<string | number>
> extends ProviderBase<ICustomOperationProviderOptions> {
  public readonly ids = new Set<string | number>();

  public pushOptions(options: ICustomOperationProviderOptions): void {
    for(const id of options.ids)
      this.ids.add(id);
  }

  public usedContexts(): Array<TRegisterEvaluationContext> {
    return [OperationClassifier]
  }

  public async provide(data: TProviderEvaluationContext): Promise<ICustomOperationProviderData<TOperationId>> {
    const result = {
      customOperations: {}
    } as ICustomOperationProviderData<TOperationId>;

    const accounts = await data.get(OperationClassifier);
    if (accounts.operationsPerType.custom_json_operation)
      for(const operation of accounts.operationsPerType.custom_json_operation) {
        if (!this.ids.has(operation.operation.id))
          continue;

        if (!result.customOperations[operation.operation.id])
          result.customOperations[operation.operation.id] = [];

        result.customOperations[operation.operation.id].push({
          operation: operation.operation,
          transaction: operation.transaction
        });
      }

    if (accounts.operationsPerType.custom_operation)
      for(const operation of accounts.operationsPerType.custom_operation) {
        if (!this.ids.has(operation.operation.id))
          continue;

        if (!result.customOperations[operation.operation.id])
          result.customOperations[operation.operation.id] = [];

        result.customOperations[operation.operation.id].push({
          operation: operation.operation,
          transaction: operation.transaction
        });
      }

    return result;
  }
}

File: python/workerbee/chain_observers/providers/custom_operation_provider.py
"""CustomOperationProvider — provides custom_json operations grouped by id.

Mirrors src/chain-observers/providers/custom-operation-provider.ts.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from .._ordered_set import OrderedSet
from ..classifiers.collector_classifier_base import TRegisterEvaluationContext
from ..classifiers.operation_classifier import OperationClassifier
from .provider_base import ProviderBase

if TYPE_CHECKING:
    from ..factories.data_evaluation_context import DataEvaluationContext
    from ..payloads import CustomOperationsPayload, OpBodyTransactionPair

CustomOperationOptions = dict[str, list[str]]


class CustomOperationProvider(ProviderBase):
    def __init__(self) -> None:
        self.ids: OrderedSet[str] = OrderedSet()

    def push_options(self, options: CustomOperationOptions) -> None:
        ids = options.get("ids", [])
        if not ids:
            raise ValueError("ids must not be empty")
        for op_id in ids:
            self.ids.add(op_id)

    def used_contexts(self) -> list[TRegisterEvaluationContext]:
        return [OperationClassifier]

    async def provide(self, data: DataEvaluationContext) -> CustomOperationsPayload:
        result: dict[str, list[OpBodyTransactionPair]] = {}

        operations = await data.get(OperationClassifier)
        custom_json_ops = operations["operations_per_type"].get("custom_json_operation")
        if custom_json_ops:
            for operation in custom_json_ops:
                op_id = operation["operation"]["id"]
                if not isinstance(op_id, str):
                    continue
                if op_id not in self.ids:
                    continue

                if op_id not in result:
                    result[op_id] = []

                result[op_id].append(
                    {
                        "operation": operation["operation"],
                        "transaction": operation["transaction"],
                    }
                )

        return {"custom_operations": result}
